### PR TITLE
Add detecting and including custom modules on sys mode

### PR DESCRIPTION
**Pull request by EPashkin**
Fix https://github.com/gtk-rs/gir/issues/512

cc @GuillaumeGomez, @sdroege

### CODE REVIEW

File: src/codegen/sys/lib_.rs [hunk 1]
```diff
@@ -90,6 +92,36 @@ fn get_extern_crate_string(library: &ExternalLibrary) -> String {
     )
 }
 
+fn include_custom_modules(w: &mut Write, env: &Env) -> Result<()> {
+    let modules = try!(find_modules(env));
+    if !modules.is_empty() {
+        try!(writeln!(w, ""));
+        for module in modules {
+            try!(writeln!(w, "pub mod {};", module));
```
Comment (sdroege) [inline src/codegen/sys/lib_.rs]: Maybe
```
mod foo;
pub use foo::*;
```
instead?

Comment (EPashkin) [reply src/codegen/sys/lib_.rs]: You right, this give better flexibility.
@GuillaumeGomez You agree?

Comment (sdroege) [reply src/codegen/sys/lib_.rs]: Should there also be something to ignore autogenerated pieces then, so they can be manually implemented?

Comment (EPashkin) [reply src/codegen/sys/lib_.rs]: [ignoring](https://github.com/gtk-rs/gir/blob/master/README.md#the-ffi-mode-toml-config) not enough?

Comment (sdroege) [reply src/codegen/sys/lib_.rs]: That's exactly what I thought of. For some reason I didn't notice that it is already implemented for FFI :)

Comment (sdroege) [reply src/codegen/sys/lib_.rs]: Ah but it can't be used for individual functions, or can it?

Comment (EPashkin) [reply src/codegen/sys/lib_.rs]: Can same as in normal version as [code](https://github.com/gtk-rs/gir/blob/master/src/codegen/sys/functions.rs#L191) still here.

Comment (GuillaumeGomez) [reply src/codegen/sys/lib_.rs]: > Maybe
>
> ```
> mod foo;
> pub use foo::*;
> ```
>
> instead?

I agree on it. :)

**Comment by EPashkin** [comment]
Updated.

File: src/codegen/sys/lib_.rs [hunk 2]
```diff
@@ -90,6 +92,40 @@ fn get_extern_crate_string(library: &ExternalLibrary) -> String {
...
+fn find_modules(env: &Env) -> Result<Vec<String>> {
+    let path = env.config.target_path.join("src");
+
+    let mut vec = Vec::<String>::new();
+    for entry in try!(fs::read_dir(path)) {
+        let path = try!(entry).path();
+        let ext = path.extension().unwrap();
+        if ext != "rs" { continue }
+        let file_stem = path.file_stem().unwrap();
+        if file_stem == "lib" { continue; }
```
Comment (GuillaumeGomez) [inline src/codegen/sys/lib_.rs]: `continue;` or `continue`? 😛

File: src/codegen/sys/lib_.rs [hunk 3]
```diff
@@ -90,6 +92,40 @@ fn get_extern_crate_string(library: &ExternalLibrary) -> String {
...
+    Ok(())
+}
+
+fn find_modules(env: &Env) -> Result<Vec<String>> {
+    let path = env.config.target_path.join("src");
+
+    let mut vec = Vec::<String>::new();
+    for entry in try!(fs::read_dir(path)) {
+        let path = try!(entry).path();
+        let ext = path.extension().unwrap();
```
Comment (GuillaumeGomez) [inline src/codegen/sys/lib_.rs]: Also, can you replace all `.unwrap()` with `.expect(...)` please? We should try to keep this as a good habit.

File: src/codegen/sys/lib_.rs (hunk 2 above)
Comment (EPashkin) [reply src/codegen/sys/lib_.rs]: Thanks, I use `cargo fmt` to this function.

File: src/codegen/sys/lib_.rs (hunk 3 above)
Comment (EPashkin) [reply src/codegen/sys/lib_.rs]: Updated but IMHO Problem with expect, that it need include "path" to be clear and `.unwrap_or_else(|| panic!(...))` take much space.

**Comment by EPashkin** [comment]
Updated

**Comment by sdroege** [comment]
👍

**Pull request merged**